ada/extensions/src/libadalang-helpers.adb

```diff
@@ -0,0 +1,122 @@
+
+         Ctx : Analysis_Context;
+
+         Files : String_Vectors.Vector;
+
+         Units : Unit_Vectors.Vector;
+
+      begin
+         --  Setup traces from config file
+         GNATCOLL.Traces.Parse_Config_File;
+
+         if not Args.Parser.Parse then
+            return;
+         end if;
+
+         Ctx := Create_Context
+           (Charset       => +Args.Charset.Get,
+            Unit_Provider => UFP);
+
+         --  Get explicitly passed files if necessary
+         if not Args.Files_From_Project.Get then
+            for F of Args.Files.Get loop
+               Files.Append (F);
+            end loop;
+         end if;
+
+         --  Handle project file
+         if Length (Args.Project_File.Get) > 0 then
+            declare
+               Filename : constant String := +Args.Project_File.Get;
+               Env      : Project_Environment_Access;
+               Project  : constant Project_Tree_Access := new Project_Tree;
+               List     : File_Array_Access;
+            begin
+               Initialize (Env);
+
+               --  Set scenario variables
+               for Assoc of Args.Scenario_Vars.Get loop
+                  declare
+                     A        : constant String := +Assoc;
+                     Eq_Index : Natural := A'First;
+                  begin
+                     while Eq_Index <= A'Length
+                       and then A (Eq_Index) /= '=' loop
+                        Eq_Index := Eq_Index + 1;
+                     end loop;
+                     if Eq_Index not in A'Range then
+                        Put_Line ("Invalid scenario variable: -X" & A);
+                        Ada.Command_Line.Set_Exit_Status
+                          (Ada.Command_Line.Failure);
+                        return;
+                     end if;
+                     Change_Environment
+                       (Env.all,
+                        A (A'First .. Eq_Index - 1),
+                        A (Eq_Index + 1 .. A'Last));
+                  end;
+               end loop;
+
+               Load (Project.all, Create (+Filename), Env);
+               UFP := Create_Project_Unit_Provider_Reference (Project, Env);
+
+               if Args.Files_From_Project.Get then
+
+                  List := Project.Root_Project.Source_Files;
+
+                  Sort (List.all);
+
+                  for F of List.all loop
+                     declare
+                        FI        : constant File_Info := Project.Info (F);
+                        Full_Name : Filesystem_String renames F.Full_Name.all;
+                        Name      : constant String := +Full_Name;
+                     begin
+                        if FI.Language = "ada" then
+                           Files.Append (+Name);
+                        end if;
+                     end;
+                  end loop;
+                  Unchecked_Free (List);
+
+               end if;
+            end;
+         end if;
```

Comment:
I think we should terminate with an error if `--files-from-project` is passed while no project file is passed.

Reply:
Right, good catch, thanks!

Reply:
Although I'm starting to wonder... If you just pass a project file and no explicit files, `files-from-project` mode will be used. So is it worth keeping the explicit command line option at all?

Reply:
As agreed live, I'll remove the option